presto-main/src/main/java/com/facebook/presto/operator/Driver.java

```diff
@@ -272,21 +284,15 @@ private void processNewSources()
             return driverBlockedFuture;
         }
 
-        ImmutableList<SettableFuture<?>> revokingRequestedFutures = operators.stream()
+        // Unblock as soon as the driver is unblocked or we get a revoking request for any of the operators
+        SettableFuture<?> result = SettableFuture.create();
+        Runnable listener = () -> result.set(null);
+        driverBlockedFuture.addListener(listener, newDirectExecutorService());
+        operators.stream()
+                .filter(operator -> !revokingOperators.containsKey(operator))
                 .map(Operator::getOperatorContext)
-                .map(OperatorContext::getMemoryRevokingRequestedFuture)
-                .collect(toImmutableList());
-        Optional<SettableFuture<?>> doneRevokingRequestedFuture = revokingRequestedFutures.stream().filter(Future::isDone).findAny();
-        if (doneRevokingRequestedFuture.isPresent()) {
-            // revoking already requested for one of operators; return completed future
-            return doneRevokingRequestedFuture.get();
-        }
-
-        // unblock as soon as driver is unblocked or we get revoking request for any of operators
-        ImmutableList.Builder<ListenableFuture<?>> futures = ImmutableList.builder();
-        futures.add(driverBlockedFuture);
-        futures.addAll(revokingRequestedFutures);
-        return firstFinishedFuture(futures.build());
+                .forEach(operatorContext -> operatorContext.setMemoryRevokingRequestedListener(listener));
```

Comment:
Why do we set a new listener on every loop of the driver? It would be cleaner for the listener to be set once when the driver is constructed and let the listener (and this class) deal with the complexities of what state needs to change on every loop, etc. It would also simplify the semantics of how setListener behaves with respect to calls to it being interleaved with requestRevocation/resetRevocation